### PR TITLE
Fix context shim override behavior

### PR DESCRIPTION
**Pull request by henrique-ft**
I'll try to summarize the problem as much as possible:

The split has this class:

```ruby
class ContextShim
  include Split::Helper
  public :ab_test, :ab_finished

  def initialize(context)
    @context = context
  end

  def ab_user
    @ab_user ||= Split::User.new(@context)
  end
end
```
Which includes:

```ruby
include Split::Helper
```
If we use the context shim this way:

```ruby
def split
  # self is a class that has the methods 'params' and 'request'
  @split ||= ::Split::EncapsulatedHelper::ContextShim.new(self) 
end
```

inside a high level abstraction, we got a problem.

### Problem
The override of ab via URL (`?ab_test[my_test]=variant`) and Split panel (cookies) doesn't work.
It doesn't work because `Split::Helper` uses these methods:

```ruby
def override_alternative_by_params(experiment_name)
  defined?(params) && params[OVERRIDE_PARAM_NAME] && params[OVERRIDE_PARAM_NAME][experiment_name]
end

def override_alternative_by_cookies(experiment_name)
  return unless defined?(request)

  if request.cookies && request.cookies.key?("split_override")
    experiments = JSON.parse(request.cookies["split_override"]) rescue {}
    experiments[experiment_name]
  end
end
```

Which depend on `defined?(params)` and `defined?(request)`, which are in `@context` but not in the scope of `ContextShim`.
In other words:
If we add params and request to the scope of `ContextShim`, we can solve the problem.

### Solution

Include these methods in `ContextShim`:

```ruby
def params
  request.params if request_present?
end

def request
  @context.request if @context.respond_to?(:request)
end
```

I also create a `request_present?` and `params_present?` methods in `Split::Helper` to avoid problems when request or params are defined but returning `nil`: This is the case when `@context` don't implement `request` or `params`

### CODE REVIEW

File: spec/encapsulated_helper_spec.rb [hunk 1]
```diff
@@ -5,19 +5,21 @@
...
-    it "should not raise an error when params raises an error" do
-      expect { params }.to raise_error(NoMethodError)
-      expect { ab_test("link_color", "blue", "red") }.not_to raise_error
+    context "when params raises an error" do
+      before do
+        allow(self).to receive(:params).and_raise(NoMethodError)
+      end
+
+      it "should not raise an error " do
+        expect { params }.to raise_error(NoMethodError)
```
[the remark this answers is not in the document]
Comment (henrique-ft) [reply spec/encapsulated_helper_spec.rb:20]: and this definitions are inside the spec

**Comment by andrehjr** [comment]
I'll merge this as-is. I'm making some changes on a new PR. Not sure why github actions didn't run on this PR 🤔 

Main branch was broken due some other issue with newer rack. I've fixed that too.

File: spec/encapsulated_helper_spec.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrehjr) [reply spec/encapsulated_helper_spec.rb:20]: Removed this on #723

[the remark this answers is not in the document]
Comment (andrehjr) [reply spec/encapsulated_helper_spec.rb:20]: Also checked if we were re-introducing https://github.com/splitrb/split/pull/257

Which was why the NoMethodError was added in first place... Tested in Rails 7+ and it was ok.
